src/context/MultiLineSelectContext/index.tsx

```diff
@@ -0,0 +1,124 @@
+import {
+  createContext,
+  useCallback,
+  useContext,
+  useMemo,
+  useState,
+} from "react";
+import { QueryParams } from "constants/queryParams";
+import useLineRangeSelection from "hooks/useLineRangeSelection";
+import { useQueryParam } from "hooks/useQueryParam";
+import { leaveBreadcrumb } from "utils/errorReporting";
+
+type MultiLineSelectContextState = {
+  handleSelectLine: (selectedLine: number, shiftClick: boolean) => void;
+  handleCloseMenu: () => void;
+  menuPosition: number | undefined;
+  openMenu: boolean;
+  selectedLines: {
+    startingLine?: number;
+    endingLine?: number;
+  };
+};
+
+const MultiLineSelectContext =
+  createContext<MultiLineSelectContextState | null>(null);
+
+const useMultiLineSelectContext = () => {
+  const context = useContext(MultiLineSelectContext);
+  if (context === undefined) {
+    throw new Error(
+      "useMultiLineSelectContext must be used within a MultiLineSelectContextProvider"
+    );
+  }
+  return context as MultiLineSelectContextState;
+};
+
+const MultiLineSelectContextProvider: React.FC<{
+  children: React.ReactNode;
+}> = ({ children }) => {
+  const [selectedLines, setSelectedLines] = useLineRangeSelection();
+  const [openMenu, setOpenMenu] = useState<boolean>(false);
+  const [shareLine] = useQueryParam<number | undefined>(
+    QueryParams.ShareLine,
+    undefined
+  );
+
+  const hasShareLine = shareLine !== undefined;
+  const hasEndingLine = selectedLines.endingLine !== undefined;
+
+  const initialMenuPosition =
+    hasShareLine && hasEndingLine
+      ? selectedLines.endingLine
+      : selectedLines.startingLine;
+
+  const [menuPosition, setMenuPosition] = useState<number | undefined>(
+    initialMenuPosition ?? undefined
+  );
+
+  const clearSelection = useCallback(() => {
```

Comment:
Right now (afaict) the only way to clear the selection is to click a few times: the first click closes the menu if it's open, the next replaces the selection range with a single selected line, and the third click clears that line.

Do you think a "clear selection" menu item would be helpful, or is that overkill? Or maybe clicking within the selection range should just clear the selection instead of replacing it with a single line selection?

Reply:
Just added a Clear selection menu item.